File: cmd/smolanalytics/connect.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

// connect wires this binary into the MCP config of the coding assistants you already use,
// so "ask your analytics in your editor" is one command instead of hand-editing JSON. It
// detects installed assistants and merges a "smolanalytics" stdio server into each config
// (preserving anything already there), and never clobbers a config it can't parse. Run it
// once, restart the editor, ask away.
//
//	smolanalytics connect            # every installed assistant
//	smolanalytics connect cursor     # just one (claude | cursor | windsurf | vscode | cline | claude-code)

// mcpClient is an assistant we can auto-configure by merging into its JSON config.
type mcpClient struct {
	name  string // display name
	short string // arg alias
	path  string // config file ("" if unknown on this OS)
	key   string // top-level object key: "mcpServers" (most) or "servers" (VS Code)
}

func connect(args []string) {
	host, key := parseHostKey(args)
	// The first bare arg is the editor — but a flag's VALUE is also bare. Scanning naively
	// meant `connect --host https://x.fly.dev --key sa_k` read the URL as the editor name,
	// matched no assistant, and silently configured nothing. That is the documented cloud
	// command, so for cloud users this subcommand did nothing at all and said so nowhere.
	target := parseConnectTarget(args)

	bin, err := os.Executable()
	if err != nil {
		fmt.Println("couldn't find this binary's path:", err)
		return
	}
	bin, _ = filepath.Abs(bin)
	data, _ := filepath.Abs(dataPath())

	// cloud mode (--host + --key): wire a local stdio proxy so the agent can instrument
	// the local repo while data queries hit the cloud instance. Otherwise point `mcp` at
	// the local data file (self-host).
	cloud := host != "" && key != ""
	var entry map[string]any
	if cloud {
		host = strings.TrimRight(host, "/")
		entry = map[string]any{"command": bin, "args": []string{"mcp", "--host", host, "--key", key}}
	} else {
		entry = map[string]any{"command": bin, "args": []string{"mcp"}, "env": map[string]string{"SMOLANALYTICS_DB": data}}
	}

	explicit := target != "" && target != "all"
	wrote, failed := 0, 0
	var skipped []mcpClient

	for _, c := range mcpClients() {
		if explicit && !strings.EqualFold(target, c.short) && !strings.EqualFold(target, c.name) {
			continue
		}
		if c.path == "" {
			continue
		}
		// on auto-detect, only touch assistants that are actually installed (config dir
		// exists). SAY when we skip one: silently passing over a client the user is sitting
		// in front of is why this felt broken — you ran connect, your assistant was not
		// mentioned at all, and there was nothing to act on.
		if !explicit && !dirExists(filepath.Dir(c.path)) {
			skipped = append(skipped, c)
			continue
		}
		if err := mergeMCPConfig(c.path, c.key, entry); err != nil {
			fmt.Printf("  ✗ %s: %v\n", c.name, err)
			failed++
			continue
		}
		fmt.Printf("  ✓ %s\n      %s\n", c.name, c.path)
		wrote++
	}

	// Claude Code configures via its own CLI (the documented path), not a file we edit.
	if (!explicit || strings.EqualFold(target, "claude-code")) && addClaudeCode(bin, data, host, key) {
		fmt.Println("  ✓ Claude Code  (added via `claude mcp add`)")
		wrote++
	}

	// Report what we did NOT touch, and exactly how to force it. A config dir only appears
	// once the app has run at least once, so "not installed" here often means "installed but
	// never opened", which is a state the user can fix in two seconds if we say so.
	if len(skipped) > 0 {
		fmt.Println()
		fmt.Println("Not configured (no config folder yet, so the app looks uninstalled):")
		for _, c := range skipped {
			fmt.Printf("  – %-16s expected %s\n", c.name, c.path)
		}
		fmt.Println()
		fmt.Println("If one of those IS installed, open it once so it creates its config, or force it:")
		fmt.Printf("  smolanalytics connect --target %s\n", skipped[0].short)
	}

	if wrote == 0 {
		fmt.Println()
		fmt.Println("Nothing was configured automatically. Add this by hand")
		fmt.Println("(most use the \"mcpServers\" key; VS Code uses \"servers\"):")
		fmt.Println()
		b, _ := json.MarshalIndent(map[string]any{"mcpServers": map[string]any{"smolanalytics": entry}}, "  ", "  ")
		fmt.Println("  " + string(b))
		return
	}
	fmt.Println()
	if failed > 0 {
		fmt.Printf("%d assistant(s) could not be written. The JSON above can be pasted by hand.\n", failed)
	}
	fmt.Println("Done. Now FULLY restart the assistant, not just its window:")
	fmt.Println("  Claude Desktop  quit with Cmd-Q (closing the window is not enough), then reopen")
	fmt.Println("  Cursor/Windsurf reload the window, or restart the app")
	fmt.Println()
	fmt.Println("Then ask: \"what's my biggest funnel drop-off this week?\"")
}

// mcpClients lists every assistant we can auto-configure, with OS-specific config paths.
func mcpClients() []mcpClient {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	var cfg string // per-OS user-config base
	switch runtime.GOOS {
	case "darwin":
		cfg = filepath.Join(home, "Library", "Application Support")
	case "windows":
		cfg = os.Getenv("APPDATA")
	default:
		cfg = filepath.Join(home, ".config")
	}
	return []mcpClient{
		{"Claude Desktop", "claude", filepath.Join(cfg, "Claude", "claude_desktop_config.json"), "mcpServers"},
		{"Cursor", "cursor", filepath.Join(home, ".cursor", "mcp.json"), "mcpServers"},
		{"Windsurf", "windsurf", filepath.Join(home, ".codeium", "windsurf", "mcp_config.json"), "mcpServers"},
		{"VS Code", "vscode", filepath.Join(cfg, "Code", "User", "mcp.json"), "servers"},
		{"Cline", "cline", filepath.Join(cfg, "Code", "User", "globalStorage", "saoudrizwan.claude-dev", "settings", "cline_mcp_settings.json"), "mcpServers"},
	}
}

// mergeMCPConfig adds (or replaces) the smolanalytics server under `key` in an assistant's
// config, keeping every other server intact. Refuses to overwrite a config it can't parse.
func mergeMCPConfig(path, key string, entry map[string]any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	cfg := map[string]any{}
	if b, err := os.ReadFile(path); err == nil && len(b) > 0 {
		if err := json.Unmarshal(b, &cfg); err != nil {
			return fmt.Errorf("existing config isn't valid JSON, leaving it untouched: %s", path)
		}
	}
	servers, ok := cfg[key].(map[string]any)
	if !ok || servers == nil {
		servers = map[string]any{}
	}
	servers["smolanalytics"] = entry
	cfg[key] = servers
	out, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, out, 0o600)
}

// addClaudeCode registers the server through the `claude` CLI (Claude Code's own way).
// Returns true if it succeeded. Best-effort: no CLI, no problem.
func addClaudeCode(bin, data, host, key string) bool {
	if _, err := exec.LookPath("claude"); err != nil {
		return false
	}
	// remove any prior entry so re-running connect is idempotent, then add.
	_ = exec.Command("claude", "mcp", "remove", "smolanalytics", "-s", "user").Run()
	var cmd *exec.Cmd
	if host != "" && key != "" { // cloud proxy
		cmd = exec.Command("claude", "mcp", "add", "smolanalytics", "-s", "user",
			"--", bin, "mcp", "--host", strings.TrimRight(host, "/"), "--key", key)
	} else {
		cmd = exec.Command("claude", "mcp", "add", "smolanalytics", "-s", "user",
			"-e", "SMOLANALYTICS_DB="+data, "--", bin, "mcp")
	}
	return cmd.Run() == nil
}

func dirExists(p string) bool {
	fi, err := os.Stat(p)
	return err == nil && fi.IsDir()
}

// flagTakesValue reports whether a connect flag consumes the following argument, so the
// editor-name scan does not mistake a flag's value for an assistant name.
func flagTakesValue(flag string) bool {
	switch strings.TrimLeft(flag, "-") {
	case "host", "key", "target", "data":
		return true
	}
	return false
}

// parseConnectTarget picks the editor name out of the args, skipping any flag values.
//
// It used to be an inline loop taking the first argument without a leading "-". A flag's
// value is also bare, so `connect --host https://x.fly.dev --key sa_k` read the URL as the
// editor, matched nothing, and configured nothing while printing no error.
func parseConnectTarget(args []string) string {
	skipNext := false
	for _, a := range args {
		if skipNext {
			skipNext = false
			continue
		}
		if strings.HasPrefix(a, "-") {
			// --flag=value carries its own value; --flag consumes the next argument
			if !strings.Contains(a, "=") && flagTakesValue(a) {
				skipNext = true
			}
			continue
		}
		return a
	}
	return ""
}

File: cli/lib/connect.mjs
// `npx smolanalytics connect [editor]` — wire the smolanalytics MCP server into every
// assistant on this machine, or one named editor.
//
// This is the npm port of the engine's `connect` subcommand, HTTP-remote flavor: the
// hosted product speaks MCP over HTTP (your org endpoint at smolanalytics.com/api/mcp,
// or a single instance's /mcp), so the entry we write is a url + bearer header, never a
// local process. The per-editor mechanics mirror the engine source exactly: most clients
// key the entry under "mcpServers", VS Code uses "servers", and Claude Code is configured
// through its own CLI (`claude mcp add`) because that is its documented path.
//
// Same house rules as init: never touch a file without saying which file and what changes,
// and never leave a half-edit behind. An unparseable existing config is REPORTED and left
// alone — clobbering someone's hand-tuned mcp.json to add ourselves is how trust ends.

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

export const DEFAULT_ENDPOINT = "https://smolanalytics.com/api/mcp";

// endpointFrom normalizes what the user typed: a bare instance host gets "/mcp" appended,
// anything already pointing at an mcp path is used verbatim, and no input means the cloud
// org endpoint.
export function endpointFrom(url) {
  const base = (url || "").trim().replace(/\/+$/, "");
  if (!base) return DEFAULT_ENDPOINT;
  if (base.endsWith("/mcp")) return base;
  return base + "/mcp";
}

// entryFor is the one remote-server shape every editor's JSON shares.
export function entryFor(endpoint, key) {
  return { type: "http", url: endpoint, headers: { Authorization: `Bearer ${key}` } };
}

// clients lists every assistant we can auto-configure, with OS-specific config paths.
// `key` is the top-level object key the editor expects.
export function clients(platform = process.platform, home = os.homedir()) {
  let cfg;
  switch (platform) {
    case "darwin":
      cfg = path.join(home, "Library", "Application Support");
      break;
    case "win32":
      cfg = process.env.APPDATA || path.join(home, "AppData", "Roaming");
      break;
    default:
      cfg = process.env.XDG_CONFIG_HOME || path.join(home, ".config");
  }
  return [
    { name: "Claude Desktop", short: "claude-desktop", path: path.join(cfg, "Claude", "claude_desktop_config.json"), key: "mcpServers" },
    { name: "Cursor", short: "cursor", path: path.join(home, ".cursor", "mcp.json"), key: "mcpServers" },
    { name: "Windsurf", short: "windsurf", path: path.join(home, ".codeium", "windsurf", "mcp_config.json"), key: "mcpServers" },
    { name: "VS Code", short: "vscode", path: path.join(cfg, "Code", "User", "mcp.json"), key: "servers" },
    { name: "Cline", short: "cline", path: path.join(cfg, "Code", "User", "globalStorage", "saoudrizwan.claude-dev", "settings", "cline_mcp_settings.json"), key: "mcpServers" },
  ];
}

// mergeConfig adds/replaces ONLY the "smolanalytics" entry under the editor's top-level
// key, preserving everything else in the file byte-for-meaning. Returns null on success,
// an error message otherwise. An existing file that does not parse is an error, not a
// license to overwrite.
export function mergeConfig(file, topKey, entry, io = fs) {
  let cfg = {};
  if (io.existsSync(file)) {
    try {
      const raw = io.readFileSync(file, "utf8");
      cfg = raw.trim() ? JSON.parse(raw) : {};
    } catch {
      return `${file} exists but is not valid JSON — fix it (or delete it) and re-run; nothing was changed`;
    }
    if (cfg === null || typeof cfg !== "object" || Array.isArray(cfg)) {
      return `${file} is not a JSON object — nothing was changed`;
    }
  }
  const servers = cfg[topKey] && typeof cfg[topKey] === "object" && !Array.isArray(cfg[topKey]) ? cfg[topKey] : {};
  servers["smolanalytics"] = entry;
  cfg[topKey] = servers;
  io.mkdirSync(path.dirname(file), { recursive: true });
  io.writeFileSync(file, JSON.stringify(cfg, null, 2) + "\n");
  return null;
}

// addClaudeCode registers through Claude Code's own CLI — the documented path — scoped to
// the user so it works in every repo. Returns true if the command ran and succeeded.
export function addClaudeCode(endpoint, key, run = spawnSync) {
  const probe = run("claude", ["--version"], { stdio: "ignore" });
  if (probe.error || probe.status !== 0) return false;
  // remove-then-add so a re-run updates the token instead of failing on "already exists"
  run("claude", ["mcp", "remove", "-s", "user", "smolanalytics"], { stdio: "ignore" });
  const res = run(
    "claude",
    ["mcp", "add", "-s", "user", "--transport", "http", "smolanalytics", endpoint, "--header", `Authorization: Bearer ${key}`],
    { stdio: "ignore" },
  );
  return !res.error && res.status === 0;
}

export function connectCmd({ url, key, target, log = console.log, io = fs, run = spawnSync, platform = process.platform }) {
  if (!key) {
    log("");
    log("connect needs your MCP token:");
    log("  npx smolanalytics connect --key sa_org_...        (org API token: smolanalytics.com → Settings)");
    log("  npx smolanalytics connect --key ... --url https://YOUR-INSTANCE   (one instance instead of the org)");
    log("");
    log("One connection operates every project: pass project=\"the-name\" to any tool.");
    return 1;
  }
  const endpoint = endpointFrom(url);
  const entry = entryFor(endpoint, key);
  const explicit = target && target !== "all";

  let wrote = 0;
  let failed = 0;
  const skipped = [];

  for (const c of clients(platform)) {
    if (explicit && target.toLowerCase() !== c.short && target.toLowerCase() !== c.name.toLowerCase()) continue;
    // on auto-detect, only touch assistants that are actually installed (their config dir
    // exists) — and SAY when we skip one, so "my editor wasn't mentioned" never happens.
    if (!explicit && !io.existsSync(path.dirname(c.path))) {
      skipped.push(c);
      continue;
    }
    const err = mergeConfig(c.path, c.key, entry, io);
    if (err) {
      log(`  x ${c.name}: ${err}`);
      failed++;
    } else {
      log(`  + ${c.name}`);
      log(`      ${c.path}`);
      wrote++;
    }
  }

  if (!explicit || target === "claude-code") {
    if (addClaudeCode(endpoint, key, run)) {
      log("  + Claude Code  (added via `claude mcp add -s user`)");
      wrote++;
    } else if (explicit) {
      log("  x Claude Code: the `claude` CLI was not found on PATH");
      failed++;
    }
  }

  if (skipped.length > 0) {
    log("");
    log("Not configured (no config folder yet, so the app looks uninstalled):");
    for (const c of skipped) log(`  - ${c.name.padEnd(16)} expected ${c.path}`);
    log("");
    log("If one of those IS installed, open it once so it creates its config, or force it:");
    log(`  npx smolanalytics connect ${skipped[0].short} --key ...`);
  }

  if (wrote === 0) {
    log("");
    log('Nothing was configured automatically. Add this by hand (VS Code uses "servers" instead of "mcpServers"):');
    log("");
    log(JSON.stringify({ mcpServers: { smolanalytics: entry } }, null, 2));
    return failed > 0 ? 1 : 0;
  }

  log("");
  log("Done. Now FULLY restart the assistant, not just its window:");
  log("  Claude Desktop   quit with Cmd-Q (closing the window is not enough), then reopen");
  log("  Cursor/Windsurf  reload the window, or restart the app");
  log("");
  log('Then ask: "investigate — what needs me today?"');
  return failed > 0 ? 1 : 0;
}
